framework/include/base/MooseObject.h

```diff
@@ -32,6 +82,8 @@ InputParameters validParams<MooseObject>();
  * Every object that can be built by the factory should be derived from this class.
  */
 class MooseObject :
+  public virtual BlockRestrictableHelper, // see BlockRestrictable.h
+  public virtual BoundaryRestrictableHelper, // see BoundaryRestrictable.h
```

Comment:
You might sit down and actually draw a design (UML style) for this refactor. Basically you have two very powerful OO tools at your disposal: templating and polymorphism we are kind of halfway using both in this refactor, but I don't really feel that we've fixed any of the deficiencies of the current design. Yes I can see that the new organization does enable the control logic to work but it doesn't feel cleaner at all (i.e. we still have a unique warehouse for each object type with code in each one)

If you start from a base class and add a few intermediate classes for the differentness classes of objects the design should be abstract enough to have a reduced set of warehouses that can handle the cases where we have restrictions, execution types, and active lists. We need to fully enumerate all of these object classes. Also what are all of the ways that we need to reach into the warehouses to pull lists of objects out? Basically I want a defensible design that's is based around functional interfaces, not organization. It's not enough to say that if feels like we should split one big class into two because it's big. It should be split because you actually use the intermediate methods at some point in the framework. 

I'm not trying to be hard on you. I just think that a huge refactor of the core internals of MOOSE deserves a first class design.